Pruning tests (#3309)
input-output-hk/ouroboros-network: docs/network-spec/connection-manager.tex

rate limits connections based on how many connections are in this state, we
could end up exceeding server hard limit.

> These are all transitions that can lead to exceeding server hard limit:

Reviewer: ```suggestion
These are all transitions that potentialy could lead to exceeding server hard limit:
```

\begin{itemize}
  \item \DuplexState{} to \InboundStateDup{} (via \DemotedToColdDupLoc{})
  \item \OutboundStateDupTau{} to \InboundStateDup{} (via \DemotedToColdDupLoc{})
  \item \OutboundStateDupTau{} to \DuplexState{} (via \PromotedToWarmDupRem{})
  \item \OutboundStateDup{} to \DuplexState{} (via \PromotedToWarmDupRem{})
\end{itemize}

To solve this problem, when a connection is transitioned from
\DuplexState{} to \InboundStateDup{} (via \DemotedToColdDupLoc{}), for example,
the \connmngr{} will check if the server hard limit was exceeded. If that
happened, the \connmngr{} will reset an arbitrary connection (with some preference).

The reason why going from \OutboundStateDupTau{} (or \OutboundStateDup{}) to
\InboundStateDup{} might exceed the server hard limit is exacty the same as the
\DuplexState{} to \InboundStateDup{} one.  However, the reason why going from

> \OutboundStateDupTau{} to \DuplexState{} might exceed the hard limit is more tricky.

Reviewer: ```suggestion
\OutboundStateDupTau{} to \DuplexState{} might exceed the limit is more tricky.
```
I think it's clear from the context which limit we are talking about.

To reach a \DuplexState{} one assumes there must have been an incoming
\textit{accepted} connection, but there's another way that two end-points
can establish a connection without a node accepting it. If two nodes try to request an
outbound connection simultaneously, it is possible, for two applications to both perform
an active open to each other at the same time.  This is called a
\href{https://flylib.com/books/en/3.223.1.190/1/}{\textit{simultaneous open}}.
In a simultaneous TCP open, we can have 2 nodes establishing a connection without any of

> them having explicitly accepted a connection, which can make a server serving more peers
> than the hard limit.

Reviewer: ```suggestion
them having explicitly accepted a connection, which can make the server violate its file descriptor limit.
```

\connmngr{} will check if the server hard limit was exceeded. If that
happened, the \connmngr{} will reset an arbitrary connection (with some preference).

There are actually two more transitions that can lead to exceeding server hard limit:
\begin{itemize}
  \item \OutboundStateDupTau{} to \InboundStateDup{} (via \DemotedToColdDupLoc{})
  \item \OutboundStateDupTau{} to \DuplexState{} (via \PromotedToWarmDupRem{})

> \end{itemize}

Reviewer: I think it's better to change thee paragraph above and itemize (as you do here) all three transitions which do pruning in one place.

The reason why going from \OutboundStateDupTau{} to \InboundStateDup{} might exceed the
server hard limit is exacty the same as the \DuplexState{} to \InboundStateDup{} one.
However, the reason why going from \OutboundStateDupTau{} to \DuplexState{} might exceed
the hard limit is more tricky. To reach a \DuplexState{} one assumes there must have been
an incoming \textit{accepted} connection, but there's another way that two end-points
can establish a connection without a node accepting it. If two nodes try to request an

> outbound connection simultaneously, it is possible, although improbable,

Reviewer: I wouldn't say it's improbable.  Karl has been observing simultaneous open connections regularly.  It's all about timing, if the two server are far enough of each other and there's some event that triggers both of them to reconnect (a previous connection failed, and they re-connect after some small delay), then it becomes quite probable in a large network.

for two applications to both perform an active open to each other at the same time.
This is called a \href{https://flylib.com/books/en/3.223.1.190/1/}{\textit{simultaneous
open}}. In a simultaneous TCP open, we can have 2 nodes establishing a connection without
any of them having explicitly accepted a connection, which can make a server serving more

> peers than the hard limit.

Reviewer: Well done!

Given this, we prefer to reset an inbound connection rather than close an outbound
connection because from a systemic point of view, outbound connections are more
valuable than inbound ones. If we keep the number of \established{} peers to
be smaller than the server hard limit, with a right policy we should never need
to reset a connection in \DuplexState{}. However, when dealing with a connection that
transitions from \OutboundStateDupTau{} to \DuplexState{}, we actually need to
make sure this connection is closed, because we have no way to know for sure
if this connection is the result of a TCP simultaneous open and there might

> not be any other connection available to prune that can make space for this one.

Reviewer: Hmm, and what about the `PromotedToWarm^{Duplex}_{Remote} : OutboundStateDup{} -> \DuplexState{}`.  It seems we should also prune then.  This could also happen in simultaneous open: first we land in `OutboundStateDup^\tau` then the remote end delays sending packets, `TimeoutExpired` kicks in, now it starts sending and we are above the hard limit.

This actually suggests that we should do pruning after an outbound connection was negotiated.  But we should make sure that if possible we reset inbound connection and only if there isn't any we reset the outbound one.

Author: Yes good catch. The code already covers that case because pruning is done in               
```
OutboundDupState connId connThread handle _expired -> do
  ....
```

pattern matching. I will add it to the docs as well!

Reviewer: good
